src/main/java/com/google/devtools/build/lib/bazel/bzlmod/Selection.java

```diff
@@ -165,16 +170,92 @@ private static SelectionGroup computeSelectionGroup(
         allowedVersionSet.ceiling(module.getVersion()));
   }
 
+  private static ImmutableMap<UnresolvedModuleKey, ImmutableList<UnresolvedModule>>
+      computeCompatibilityLevelRangeDeps(
+          ImmutableMap<UnresolvedModuleKey, UnresolvedModule> depGraph,
+          ImmutableMap<ModuleKey, UnresolvedModule> unresolvedModules,
+          ImmutableMap<ModuleKey, SelectionGroup> selectionGroups,
+          Map<SelectionGroup, Version> selectedVersions) {
+    Map<ModuleNameAndCompatibilityLevel, Version> minimumVersions = new HashMap<>();
+    for (UnresolvedModule module : unresolvedModules.values()) {
+      Version keyVersion = module.getKey().getVersion();
+      if (keyVersion.isEmpty()) {
+        continue;
+      }
+      minimumVersions.merge(
+          ModuleNameAndCompatibilityLevel.create(module.getName(), module.getCompatibilityLevel()),
+          keyVersion,
+          Comparators::min);
+    }
+
+    Map<String, ImmutableSortedMap.Builder<Integer, UnresolvedModule>> xxx = new HashMap<>();
+    for (Map.Entry<ModuleNameAndCompatibilityLevel, Version> entry : minimumVersions.entrySet()) {
+      ModuleNameAndCompatibilityLevel moduleNameAndCompatibilityLevel = entry.getKey();
+      String moduleName = moduleNameAndCompatibilityLevel.getModuleName();
+      ModuleKey minLevelModuleKey = ModuleKey.create(moduleName, entry.getValue());
+      Version selectedVersion = selectedVersions.get(selectionGroups.get(minLevelModuleKey));
+      ImmutableSortedMap.Builder<Integer, UnresolvedModule> yyy =
+          xxx.computeIfAbsent(
+              moduleName, k -> ImmutableSortedMap.<Integer, UnresolvedModule>naturalOrder());
+      yyy.put(
+          moduleNameAndCompatibilityLevel.getCompatibilityLevel(),
+          unresolvedModules.get(ModuleKey.create(moduleName, selectedVersion)));
+    }
+    Map<String, ImmutableSortedMap<Integer, UnresolvedModule>> zzz =
+        Maps.transformValues(xxx, ImmutableSortedMap.Builder::build);
+
+    ImmutableMap.Builder<UnresolvedModuleKey, ImmutableList<UnresolvedModule>>
+        compatibilityLevelRangeModules = ImmutableMap.builder();
+    for (Map.Entry<UnresolvedModuleKey, UnresolvedModule> entry : depGraph.entrySet()) {
+      UnresolvedModuleKey key = entry.getKey();
+      int maxCompatibilityLevel = key.getMaxCompatibilityLevel();
+      if (maxCompatibilityLevel > 0) {
+        UnresolvedModule module = entry.getValue();
+        int minCompatibilityLevel = module.getCompatibilityLevel();
+        if (maxCompatibilityLevel > minCompatibilityLevel) {
+          ImmutableCollection<UnresolvedModule> extraModules =
+              zzz.get(module.getName())
+                  .subMap(
```

Comment:
TODO: Add a test for this. I had incorrect range logic here before. Need to verify it doesn't select too high or too low of compatibility modules. Also need to verify that it correctly selects minimum version and upgrades then accordingly.